### PR TITLE
User must agree to Terms and Conditions to access a charm.

### PR DESCRIPTION
**Pull request by alesstimec**

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1167/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1168/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1169/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1170/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1171/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1172/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1201/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1202/
Test FAILed.

**Comment by tasdomas** [comment]
LGTM

File: server.go [hunk 1]
```diff
@@ -60,6 +60,8 @@ type ServerParams struct {
 	// If it is empty, IdentityURL+"/v1/discharger" will be used.
 	IdentityLocation string
 
+	TermsLocation string
```
Comment (mattyw) [inline server.go:66]: Should have a comment above

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1203/
Test FAILed.

**Comment by cmars** [comment]
Almost LGTM, just need to deny access to charms with terms if the terms service isn't configured -- just so that if the service configuration gets botched in production, we don't accidentally start giving out charms without getting agreement.

File: internal/v4/archive_test.go [hunk 2]
```diff
@@ -38,12 +41,85 @@ import (
...
+	err := s.store.SetPerms(&id.URL, "read", params.Everyone, id.URL.User)
+	c.Assert(err, gc.IsNil)
+
+	archiveUrl := storeURL("~charmers/precise/terms-0/archive")
+	rec := httptesting.DoRequest(c, httptesting.DoRequestParams{
+		Handler: s.srv,
+		URL:     archiveUrl,
+		Do:      bakeryDo(client),
+	})
+	c.Assert(rec.Code, gc.Equals, http.StatusInternalServerError)
```
Comment (cmars) [inline internal/v4/archive_test.go:72]: Perfect.

**Comment by cmars** [comment]
LGTM!

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1205/
Test FAILed.

**Pull request merged**